MS Transferor code re-factoring and clean-up according to proposal specs

vkuznet opened the pull request.

vkuznet committed:
Re-factor code according to proposal specs, work on naming convention, record structure, code flow

vkuznet commented:
Code re-factoring and clean-up of intermediate changes from #9140

amaltaro commented:
@vkuznet please have a look at the report html provided by the jenkins setup and try to clean some of those (usually errors and warnings deserve attention).

Messages like
` W1201, line 402 in requestsInfo_old: Specify string format arguments as logging function parameters`

means that your logging line should be written as:
```
logger.debug("request: %s, campaignConfig: %s", reqName, campaignConfig)
```

instead of
```
logger.debug("request: %s, campaignConfig: %s" % (reqName, campaignConfig))
```

Please also describe these changes in the first comment box.

vkuznet commented:
Alan, since you merged #9217  it created a conflict with this PR. How do you want to resolve it? I already expressed my opinion in PR#9217.

amaltaro commented:
Can you please rebase and resolve this conflict?

vkuznet commented:
Alan, can you trigger Pylint tests again, it reports suggestions for code which I already fixed/no longer exists.

amaltaro commented:
@vkuznet I haven't reviewed this pull request yet, but can you squash your changes into 2 commits please (1 for src and 1 for test)? just in case I get it merged by tomorrow morning.

vkuznet commented:
@amaltaro , Alan I tested this PR in my VM and it works. The entire data flow works correctly ,i.e. MS Transferor and MSTransferorMonit. I also cleaned up all pylints errors and cleaned up the code base itself. I put in place the unified version which I had originally and it works too, i.e. it fetches aux info for given request and produce information which we can rely on, e.g. number of events, lumis, sites, etc. Once this will be merged we can start working on individual PRs to tune-up unified logic, do transfers, etc. Feel free to make another review round and you may include this PR into June release.

vkuznet commented:
and, all commits a squashed and I used internally the record structures (option C) from your gist.

amaltaro commented:
Awesome, thanks!
For future cases, please keep pylint + unit tests fixes in a separate commit, it makes it easier to review.

I've have made some deployment changes here: https://github.com/dmwm/deployment/pull/776
I still have to test and see whether we're missing anything deployment-wise.

amaltaro commented on src/python/WMCore/MicroService/Unified/Transferor.py:
```diff
@@ -131,37 +67,42 @@ def monit(self, req_status='staging'):
                 # get campaing and unified configuration
-                campain = self.requestCampaign(req)
-                conf = self.requestConfiguration(req)
+                campaign = self.requestCampaign(reqName)
```
a cache needs to be put in place here. 10min should suffice.

amaltaro commented on src/python/WMCore/MicroService/Unified/Transferor.py:
```diff
@@ -170,229 +111,169 @@ def transferor(self, req_status='assigned'):
+            # get requests from ReqMgr2 data-service for given statue
+            requestSpecs = self.svc.reqmgr.getRequestByStatus([reqStatus], detail=True)
+            requestRecords = [requestRecord(r, reqStatus) for r in requestSpecs]
+            self.logger.debug('### monit found %s requests in %s state', len(requestRecords), reqStatus)
```
monit or transferor?

amaltaro commented on src/python/WMCore/MicroService/Unified/Transferor.py:
```diff
@@ -170,229 +111,169 @@ def transferor(self, req_status='assigned'):
         if datasets and sites:
-	    subscription = PhEDExSubscription(datasets, sites, self.group)
-	    self.logger.debug("### add subscription %s for request %s" % (subscription, request))
+            subscription = PhEDExSubscription(datasets, sites, self.group)
```
Are we able to subscribe blocks as well? Or only datasets?

amaltaro commented on src/python/WMCore/MicroService/Unified/Transferor.py:
```diff
@@ -170,229 +111,169 @@ def transferor(self, req_status='assigned'):
-			      "parent_dset_1": ["list of transfer ids"]},
-		  "secondary": {"PU_dset_1": ["list of transfer ids"],
-				"PU_dset_2": ["list of transfer ids"]},
+        {"wf_A": {"timestamp": 0000
```
this data structure is no longer going to be used. We decided to adopt option C from the gist.

amaltaro reviewed (commented):
Valentin, I have put several comments in place. If you decide to address them before I wake up tomorrow, please do so in a second commit. Otherwise, we can likely proceed with this PR and then apply some fine tuning in a later stage (or I can get it in tomorrow morning).

vkuznet commented:
Alan,
these are very good comments and they show that I'm lack of some knowledge of
DMWM. I don't have time tonight and I think it should be better of if I'll
address your suggestions separately as I don't want to rush into them in a
morning and hold you off for making a release cut.
Valentin.

On  0, Alan Malta Rodrigues <notifications@github.com> wrote:
> amaltaro commented on this pull request.
> 
> Valentin, I have put several comments in place. If you decide to address them before I wake up tomorrow, please do so in a second commit. Otherwise, we can likely proceed with this PR and then apply some fine tuning in a later stage (or I can get it in tomorrow morning).
> 
> > @@ -32,6 +32,31 @@ def findParent(dataset):
>      return [str(i['parent_dataset']) for i in json.loads(data)]
>  
>  
> +def requestRecord(data, reqStatus):
> +    "Return valid fields from reqmgr2 record which we'll use in MS transferor"
> +    siteWhiteList = data.get('SiteWhitelist', [])
> +    siteBlackList = data.get('SiteBlacklist', [])
> +    tasks = [k for k in data.keys() if k.startswith('Task') and not k.endswith('Chain')]
> 
> This takes into account only TaskChain workflows. We'll have to extend it to other request types.
> 
> > @@ -32,6 +32,31 @@ def findParent(dataset):
>      return [str(i['parent_dataset']) for i in json.loads(data)]
>  
>  
> +def requestRecord(data, reqStatus):
> +    "Return valid fields from reqmgr2 record which we'll use in MS transferor"
> +    siteWhiteList = data.get('SiteWhitelist', [])
> +    siteBlackList = data.get('SiteBlacklist', [])
> +    tasks = [k for k in data.keys() if k.startswith('Task') and not k.endswith('Chain')]
> +    datasets = []
> +    for task in tasks:
> +        for key in ['InputDataset', 'MCPileup']:
> 
> It's missing `DataPileup`.
> 
> > @@ -32,6 +32,31 @@ def findParent(dataset):
>      return [str(i['parent_dataset']) for i in json.loads(data)]
>  
>  
> +def requestRecord(data, reqStatus):
> +    "Return valid fields from reqmgr2 record which we'll use in MS transferor"
> +    siteWhiteList = data.get('SiteWhitelist', [])
> +    siteBlackList = data.get('SiteBlacklist', [])
> +    tasks = [k for k in data.keys() if k.startswith('Task') and not k.endswith('Chain')]
> +    datasets = []
> +    for task in tasks:
> +        for key in ['InputDataset', 'MCPileup']:
> +            dataset = data[task].get(key, '')
> +            if dataset:
> +                datasets.append({'type': key, 'name': dataset})
> +    name = data.get('_id', '')
> 
> it should be `RequestName` key.
> 
> > @@ -32,6 +32,31 @@ def findParent(dataset):
>      return [str(i['parent_dataset']) for i in json.loads(data)]
>  
>  
> +def requestRecord(data, reqStatus):
> 
> `reqStatus` isn't used anywhere in this method, we can probably dump it.
> 
> > @@ -199,8 +224,9 @@ def getAllTasks(request, select=None, reqSpecs=None):
>  
>  def getWorkTasks(request, reqSpecs=None):
>      "Return work tasks for given request"
> -    return getAllTasks(request, select={'taskType': ['Production', 'Processing', 'Skim']}, \
> -                       reqSpecs=reqSpecs)
> +    return getAllTasks(request,
> 
> Still not sure what it does, but it deserves attention...
> 
> >  import traceback
>  
>  from httplib import HTTPException
>  
>  # WMCore modules
>  from WMCore.MicroService.Unified.Common import uConfig
> -from WMCore.MicroService.Unified.RequestInfo import requestsInfo
> -from WMCore.MicroService.Unified.TaskManager import start_new_thread, TaskManager
> +from WMCore.MicroService.Unified.RequestInfo import requestsInfo, requestRecord
> +from WMCore.MicroService.Unified.TaskManager import start_new_thread
> 
> `start_new_thread` should be renamed to `startNewThread` or alike
> 
> > -        data = {}
> -        req = str(request)
> -        if request: # check for None
> -            if req in self.db:
> -                data = self.db[req]
> -        else:
> -            for key in self.db.keys():
> -                data.update({key: self.db[key]})
> -        return data
> -
> -    def close(self):
> -        "Close backend store"
> -        self.db.close()
> -
> -def daemon(func, req_status, interval):
> +def daemon(func, reqStatus, interval, logger):
> 
> what's this `reqStatus`? A string set to `assigned`?
> 
> >          except Exception as exc:
> -            pass
> +            logger.error("MS daemon error: %s", str(exc))
> 
> given that we have no clue of what was the exception, let's make sure to record the traceback as well, so you can use `logger.exception()` instead.
> 
> > @@ -131,37 +67,42 @@ def monit(self, req_status='staging'):
>          """
>          try:
>              # get requests from ReqMgr2 data-service for given statue
> -            requestSpecs = self.svc.reqmgr.getRequestByStatus([req_status], detail=False)
> -            requests = [r for item in requestSpecs for r in item.keys()]
> -            self.logger.debug('+++ monit found %s requests in %s state' % (len(requests), req_status))
> +            # here with detail=False we get back list of records
> +            requests = self.svc.reqmgr.getRequestByStatus([reqStatus], detail=False)
> 
> This call to ReqMgr2 is actually not needed. The transfer monitoring should be performed based only on:
> * records in couchdb (containing the transfer id)
> * unified (and very likely campaign) configuration, which will hold the completion rate that we can move the request to the next status
> 
> PS.: IF, and this is an IF, we want to make sure we can handle it even if someone else does this request transition manually, then we might indeed have to query reqmgr2 to see at which status that request is.
> 
> >                  # get transfer IDs
>                  tids = self.getTransferIDs()
>                  # get transfer status
>                  transferStatuses = self.getTransferStatuses(tids)
>                  # get campaing and unified configuration
> -                campain = self.requestCampaign(req)
> -                conf = self.requestConfiguration(req)
> +                campaign = self.requestCampaign(reqName)
> 
> a cache needs to be put in place here. 10min should suffice.
> 
> > @@ -170,229 +111,169 @@ def transferor(self, req_status='assigned'):
>          https://github.com/dmwm/WMCore/wiki/ReqMgr2-MicroService-Transferor
>          """
>          try:
> +            # get requests from ReqMgr2 data-service for given statue
> +            requestSpecs = self.svc.reqmgr.getRequestByStatus([reqStatus], detail=True)
> +            requestRecords = [requestRecord(r, reqStatus) for r in requestSpecs]
> +            self.logger.debug('### monit found %s requests in %s state', len(requestRecords), reqStatus)
> 
> monit or transferor?
> 
> >          if datasets and sites:
> -	    subscription = PhEDExSubscription(datasets, sites, self.group)
> -	    self.logger.debug("### add subscription %s for request %s" % (subscription, request))
> +            subscription = PhEDExSubscription(datasets, sites, self.group)
> 
> Are we able to subscribe blocks as well? Or only datasets?
> 
> >              return tid
>  
>      def getTransferIDsDoc(self):
>          """
>          Get transfer ids document from backend. The document has the following form:
>          https://gist.github.com/amaltaro/72599f995b37a6e33566f3c749143154
> -	{"wf_A": {"timestamp": 0000
> -		  "primary": {"dset_1": ["list of transfer ids"]},
> -		  "secondary": {"PU_dset_1": ["list of transfer ids"]},
> -	 "wf_B": {"timestamp": 0000
> -		  "primary": {"dset_1": ["list of transfer ids"],
> -			      "parent_dset_1": ["list of transfer ids"]},
> -		  "secondary": {"PU_dset_1": ["list of transfer ids"],
> -				"PU_dset_2": ["list of transfer ids"]},
> +        {"wf_A": {"timestamp": 0000
> 
> this data structure is no longer going to be used. We decided to adopt option C from the gist.
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/dmwm/WMCore/pull/9216#pullrequestreview-242383769
